### ISSUE TITLE
LibTidy does not support all entities! It should!!

### ISSUE THREAD

**Issue by geoffmcl** [htacg/tidy-html5#643]
At present `libtidy`, in `entities.c`, supports only about some 253, **case sensitive**, entities... it uses `TY_(tmbstrcmp)` in the `entitiesLookup` static function...

Yet W3C pages like these - https://dev.w3.org/html5/html-author/charref - or - https://www.w3.org/TR/xml-entity-names/byalpha.html - and probably others, indicate in excess of **2,000** entities, some like `&amp;`, `&AMP;`, ... in multiple case forms... not yet exactly counted, and there may be **more**, yet to be fully explored...

Here is a live web page - https://support.apple.com/en-us/HT202021 - containing the likes of `&Tab;`, `&NewLine;`, pointed out to me by Karl Dahlke, of the [edbrowse](http://edbrowse.org/) project, which uses `libtidy` to load and `view` web pages - thanks Karl...

So at present, `edbrowse` sees warnings about **unknown** entities, and unless `preserve-entities: yes` is added to the config, they will be rendered/returned as text `&amp;Tab`, `&amp;NewLine`, along with the warnings... **not good**!

I propose `libtidy` support **all** known entities, and seek feedback, and more, other W3C references, on this... thanks...

**Comment by geoffmcl** [comment]
Have push a new branch `issue-643` which contains an `entities.c`, with some 2031 entities, scraped from the [charref](https://dev.w3.org/html5/html-author/charref) page with some quick perl...

Now processing that downloaded web page [HT202021](https://support.apple.com/en-us/HT202021) using my new tidy 5.5.76.I643, and adding `--drop-empty-elements no`, the warnings fell from 104 to just 18, and of course none of the warnings were about **unknown** entities.

And the test file pushed to my repo, [in_643.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_643.html), goes from 16 warnings, including 2 spurious `missing <li>`, to clean... 

As noted in the source, I have removed the column alignment to keep the total size down to 80K, rather than the nearly 150K for adding sort of meaningless alignment. This is of course much larger than the original 13K.

I hope this is kept sane... it does break a little from tidy's usual **good** code, especially table alignment...

And as also noted in the source, I have added some more small optimization steps to the `entitiesLookup` service which depend upon the entities table being kept in alphabetic order, and think this should be fast enough for now, rather than adding to the lib setup time by creating some type of hash. 

Also previously the entities were grouped, but not sure why... did not seem necessary...

Also I do not specifically know the html `version` that should be used on these new entities. 

For the 253 that existed, I used the version they had. But for all the **new** have used `VERS_CHECK`, which I define as `VERS_ALL`. But if some of these entities do really belong to specific html versions, then this should be replaced as appropriate...

As noted above, I have only used those in the `charref` W3 page, and not cross referenced this with other lists of entities... but it is now certainly sufficient to address this issue... but maybe there are even more...

If others, especially `edbrowse`, could checkout this `issue-643` branch and test, this could be merged to `next`, maybe in time for the upcoming 5.6 release, hopefully later this month... thanks...

**Comment by geoffmcl** [comment]
Branch `issue-643` now **re based** to `next` to add the **fixed** header install location, see issue #646, PR #647 ...

Testers must do a **pull** to update their local repo, `issue-643` branch, for further testing, **and** should make sure any installs to `include/tidy` are **deleted**... thanks...

**Comment by balthisar** [comment]
@geoffmcl, nice work. I've added some comments over in the PR.

**Comment by geoffmcl** [comment]
@balthisar thanks for your comment in the PR #644...

I have now had a chance to run my test sample [in_643.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_643.html) through the W3C validator, and it **passes**.

So then I added a [in_643-1.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_643-1.html), using a legacy `4.01` doctype, and it **FAILED**...

So I think you are right, maybe `VERS_CHECK` should be set to only html5, or **something**, until we know better... but...

I do not think a browser test is a valid test of this... 

If I was a browser writer, I think I would simply support **all** known entities, and **not** even read, or change anything, according to the doctype... they are **not** in the business of caring about **validation**... just render the document as well as possible... why not support **all** for **any** doctype...

The wikpedia [citation](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references) adds a **Standard** column which seems to clearly set a html version for each, but then does **not** seem to include the two that started this issue, namely `&Tab;` and `&NewLine;`...

And that article cites various other references, like [html40](https://www.w3.org/TR/1998/REC-html40-19980424/sgml/entities.html), which does indeed suggest many of these entities were allowed in html4 documents, in that case... Then another reference, [html32](https://www.w3.org/TR/REC-html32#latin1) which again shows a list greatly expanded on the some 252? only that **were** in tidy...

So I do think my defining `VERS_CHECK` to equal `VERS_ALL` is **not** correct, and suggest at this time it be defined as `VERS_FROM40`, which I think would get **many** of them nearly right...

Until such time someone is willing to take the time and **big** effort and set the actual version for each... and then, where possible **validate** the choice... **not** a simple task...

And I have not yet had the chance to run the regression tests on this... but can understand there may be some changes... thanks for starting to look at this...

Of course, Karl, of edbrowse, has reported this `issue-643` branch, as is, works for them... but again they just want libtidy to parse the document, and probably do not have much interest in the doctype...

What do you, and others, think about using `VERS_FROM40` for now? Thanks...

**Comment by geoffmcl** [comment]
Further exploration and understandings...

First, I **implied** the html 3 and 4 versions had a considerably **enhanced** list of entities, but seems I did not exactly count and compare them, which I have **now** done, and **no**!, it seems up to html 4 there were just some **253** known entities, which tidy has **always** supported...

And the version indicated in the [wikpedia](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references) page largely **agrees** with what tidy currently uses for these... still to tie down the minimal differences, and try to decide which is **right**, tidy or wikipedia...

It was only in **html 5** that the list jumped upwards to about 2031 entities, so those not already in tidy should **all** be marked as **`VERS_HTML5`** only. These came from the [charref](https://dev.w3.org/html5/html-author/charref) file...

Then I found this slightly bigger [html 5](https://www.w3.org/TR/html5/syntax.html#named-character-references) list, which is available as a [json](https://www.w3.org/TR/html5/entities.json) file... and note the change from calling them `entities`, to `Named character references`... wonder why, but seems ok...

This adds about an additional 94, and the only reason I can see that these did **not** make it into the `charref` list is that most, maybe all?, have more than one `codepoint`, namely 2 different `codepoints`!. 

So to add these to the tidy supported list, we can only add one or the other value, since our table can **not** have 2 values, and even if the 2 were added separately, our search and lookup services would only ever find the first value, so maybe no point... but this has not been fully checked, and to decide which is the **best** value if using only one...

Also, I have found a small number of cases where the various tables do **not** agree on even a single codepoint... Need to dig deeper here to see if I can discover the reason, and which is correct! Two known are entity `lang`, code 9001 vs 10216, and `rang`, code 9002 vs 10217... there are about 10 or so, but that seems a separate issue, after we have the bulk of the support included...

Anyway, at this moment working on adding these 94 or so, and as indicated defining `VERS_CHECK` to be `VERS_HTML5`, but also still to explore XML versions...

Also, at the same time as generating the C table structure for tidy, also generated some HTML, and it seems my chrome browser display a `glyph` for very nearly **all** - great to see... seems independent of `doctype`... will get around to pushing this/these somewhere...

So moving this forward **slowly**, and hopefully **carefully**... further feedback **very** welcome... thanks...

**Comment by geoffmcl** [comment]
Wow, it is really interesting what you discover by outputting a little HTML ;=))

I can now see why the **94**, with what I thought had **two** numeric values, are **not** in `charref`! It is **not** two values, but an `entity` that sort of **combines** two values...

Take the first, `&NotEqualTilde`. It is a **single** `glyph`, sort of combining two 'glyphs', value `&#8770;`, with `&#824;`, that `overlay` each other, so like tidy, `charref` has no current way to represent this as a **single** value!

That is they are `entities` or `named character references` that do **not** have a **single** numeric value...

Have to give some thought about how to support this in tidy... feedback welcome... thanks...

**Comment by geoffmcl** [comment]
#### Current Tidy

If you want to `view` the 253 entities currently supported by Tidy, an uploaded html display is here, [temptidyents.html](http://geoffair.org/tmp/temptidyents.html).

And have uploaded a number of other `temp` html pages... see below...

#### Two value entities

Still pondering how to make tidy **support** this **94**, or maybe **93**, entities that consist of two(2) values. Not sure what happened to the 94th? Maybe I miss counted before...

You can view these `93` here, [tempmultents.html](http://geoffair.org/tmp/tempmultents.html)... 

You can see how each is actually a combined `glyph`, made up of 2 values... sort of overlaying each other, at least in **most** cases...

And the last column shows that if you add the two values one after another they sort of work as a sort a surrogate pair, and do seem to produce the **same** single `glyph`. 

That is, for example `&NotEqualTilde;` looks the **same** as `&#8770;&#824;`, so maybe this gives a **clue** on how they **could** be supported in tidy...

As stated still pondering, exploring, this...

#### Comparison of tables

Have now looked **deeper** into the **9** differences, plus 1 missed from first, and 2 missed from the second.

A html visual is uploaded here, [tempdiffents.html](http://geoffair.org/tmp/tempdiffents.html). Some are slight **variations**, and some appear the **same**, or **equivalent**...

This is from a comparison of the 2031 entities from [charref](http://geoffair.org/tmp/tempents.c.html) page, [reference]((https://dev.w3.org/html5/html-author/charref), and the 2032 entities from the [json](http://geoffair.org/tmp/tempents2.c.html) page, [reference](https://www.w3.org/TR/html5/entities.json)...

In that table I show the `entity`, then each value as a numeric entity. So, for example, the first, `&OverBar;` shows a glyph `‾`, and `&#175;` shows `¯`, and `&#8254` shows `‾`, if these appear ok in this github markdown. So in this case the second value 8254 appears more like what the `entity` shows, so I guess that should be the value... and it seems other `choices` of the best value can be made...

And others like `&ohm;` looks to really have two possible values, 8486 or 937. How to choose between these?

#### Some conclusions

Perhaps need a second table to support the **93** that are like surrogate pairs... seek ideas, feedback on these... maybe could even be a separate issue... thanks...

It does seem the branch `issue-643`, taken from the `charref` page is **very** close... perhaps 2 or 3 could be **adjusted**, and 1 added, considering the above `difference` display, but is already largely fully **correct**, for all the single value `entities`...

So, aside from redefining `VERS_CHECK` to `VERS_HTML5`, seems PR #644 is good to go, and will merge this shortly... comments welcome... thanks...

**Comment by balthisar** [comment]
Wow, @geoffmcl, this has been a lot of work. 

In cases where they aren't acting in concert with each other, or maybe in all cases, I'd suggest the values from (the W3 reference)[https://dev.w3.org/html5/html-author/charref], at least where they look identical. 

As for how they should be internally represented, I'm not certain how Tidy handles this internally. If it encounters a named entity, is it converted to a byte? Or is it just passed through as text? I suppose that might be dependent upon `output-encoding` and `numeric-entities`, and maybe `ncr`, although in the latter case going from NCR to entity should be easy.

In any case, I'm in favor of the merge as-is, deferring the answers until later.

**Comment by geoffmcl** [comment]
@balthisar seems we sort of cross posted... 

After `visually` examining the value displays carefully, have now chosen one value over another... and made some other changes, after carefully comparing `charref` and the `json`, so think I have now **answered** most of my own questions...

Presently headed out for a while, but will write this up more on my return later... certainly now consider the PR, with 2032 known, single value, entities very complete and accurate... 

This of course does not yet address the **94** - yes found the other one I think - which are like surrogate pairs... anyway, more later...

**Comment by geoffmcl** [comment]
@balthisar as suggested above I have now **visually** examined all the differences between the two **references**, [charref](https://dev.w3.org/html5/html-author/charref) **and** [json](https://www.w3.org/TR/html5/entities.json). They are the **SAME** for some 2021 entities, and as mentioned the **differences** are enumerated here [tempdiffents.html](http://geoffair.org/tmp/tempdiffents.html). 

Note in the last two the `Name` and `G` columns have been correct! That was just a code error in the perl script, `tidyentities.pl`, used to generate that page... it previously erroniously showed `&zwnj;` for both, but their correct names are `&bsolhsub;` and `&suphsol;` as now shown...

That **visual** examination led me to the 12 changes made in commit 54f14ff9fc. 9 changes in value, removal of 1, `&race;` since it is one of the **93** special pairs, and the addition of 2, `&bsolhsub;` and `&suphsol;`, for some reason strangely missing from the `charref` reference.

For most of the changed values, the choice is simple, except perhaps `&angst;` although the `glyph` for 197 does look slightly better?, and `&ohm;`, which does seem to have 2 exactly equivalent values.

These changes were also effected in the perl script, except for the additions, so now the difference between the list generated for the `charref` reference, [tempents1.c.html](http://geoffair.org/tmp/tempents1.c.html), and the list generated for the `json` reference, [tempents12.c.html](http://geoffair.org/tmp/tempents12.c.html) is **nothing**, except for the 2 missed, namely [tempdiffents1.html](http://geoffair.org/tmp/tempdiffents1.html).

In other words, tidy's `entities.c` **agrees** 100% to the `json` reference, **except** for the **93** - yes, back to **93** since the extra one I thought I had found, `race` was already there - see [tempmultents1.html](http://geoffair.org/tmp/tempmultents1.html), which is unchanged from the previous list...

I too have yet to examine **all** this, at least for the entities with a single numeric value, in libtidy code terms, that is `how Tidy handles this internally`...

As you suggest, I think when an alphabetic, or `ascii` entity is encountered, it is `looked` up, and if found, only the codepoint is stored in the lexer - **I think**! Of couse, if it is `numeric`, either as a decimal, or hex value, then again the codepoint is stored...

Then we have quite a set of `entity` options, [MarkupEntitiesReference](http://api.html-tidy.org/tidy/quickref_next.html#MarkupEntitiesReference). Namely, with their default - `ascii-chars` **no**, `ncr` **yes**, `numeric-entities` **no**, `preserve-entities` **no**, `quote-ampersand` **yes**, `quote-marks` **no**, `quote-nbsp` **yes**, and hopefully **no** others...

So I am still to explore how each of these options effect, either the **input and decode**, `or` the **output**. I certainly feel some of the `descriptions` leave a lot to be desired...

But the extension of this `entities` list does effect two **regression** tests, namely 412 and 437468, that you mentioned... 

Was really puzzled by 412 until I dumped it and find the `-` in the title is `E2 80 93`. Current libtidy converts that to `E2 26 23 38 33 36 34 3B 26 23 38 32 32 30 3B`, ie `0xE2&#8364;&#8220;`, which seems **meaningless**!?!?. 

And the **new** tidy I643 finds the value `8220` matches `&OpenCurlyDoubleQuote;` so substitues that... If it searched more it would find 8220 also matches `&ldquo;`, but the reverse `lookup` will always return just the first found...

**Still trying to get my head around what is happening here???** 

The document/doctype is html 5, so utf-8 assumed. `E2 80 93` seems a **valid** utf-8 sequence, at least my chkutf8 app says so... 

Why do **both** current and new tidy decide to output numeric entities, especially after leaving the `E2` there??? **This is CRAZY**!!! This seems a **bug**, but maybe unrelated to this `entities` issue...

The test 437468 seems simpler, but maybe **not**

Inputs of `&ldquo;` and `&rdquo;` were output in **new** tidy as `&#8220;` and `&#8221;`, in a HTML 4.01 document/doctype...

Hmmm, not particularly **wrong**, but why, why, why??? And why were not `&OpenCurlyDoubleQuote;` and `&CloseCurlyDoubleQuote;` found? Some questions here...

While I am **not** particularly sure either of these **regression** test changes neccessarily call into question the PR #644, there does seem some questions to answer...

What are your further thoughts, or the comments of others... thanks...

**Comment by geoffmcl** [comment]
Ok, had the thought after I posted the above **regression** test questions - maybe the default config sets the character encoding, and **yes**, there it is in `config_default.conf`, **`char-encoding: latin1`**!

So now, for case 412 it becomes - What is, I think, utf-8, `E2 80 93`, for a dash, `-`, doing in a supposed `latin1` encoded file? That should be **removed**!

But still, even though that should **not** be there, and should be **removed**, that still does **not** answer **why** libtidy, **old** and **new**, converts the `80 93` to `&#8364`, which is a Euro sign, and `&#8221`, which is a Close-Curly-Double-Quote sign, or `&rdquo;`?

Tracing through the code, I can see, if in `latin1` encoding, the `80` falls into an `if ( 127 < c && c < 160 )` block, then the service `DecodeWin1252` returns the Unicode `0x20AC`, ie decimal 8364, which is the Unicode for a Euro sign, and reports an `INVALID_SGML_CHARS`, and of course that is converted to utf-8, `E2 82 AC`, and that is what is stored in the `lexer`.

Likewise, the `93` is converted to Unicode `201C`, 8221, warning reported, and stored in `lexer` as utf-8 `E2 80 9C`, so no mystery here... This action all seems correct...

**Solution:** Propose removing the utf-8 from a supposed to be `latin1` encoded file, and adjust the test  `expects` accordingly. Seems not necessary to have these additional encoding warnings in a test that is **not** related to encoding...

As usual, headed out for a while, but on return, will look deeper into the 437468...

**Comment by geoffmcl** [comment]
But before checking 437468 got curious about the `output` side of all this... and think I have discovered two **problems** now that we have this extended entity table...

But first, I had **not** noted, that under `latin1` decode the first character of the utf-8 sequence, namely `E2` got stored in the `lexer` as 2-byte utf-8 `c3 a2`... 

In **output** `PPrintText`, on encountering a char above 0x7f, it uses `TY_(GetUTF8)`, which then uses `TY_(DecodeUTF8BytesToChar)` to get the codepoint...

So encountering `c3` signals it is a 2 byte sequence, so decodes this back to codepoint `e2`, which is just sent to `AddChar`...

The next `E2 82 AC`, where the `E2` signals a 3 byte sequence, so decodes this back to codepoint `20ac`, 8364, and `PPrintChar` find this a multi-byte char, ie `(c > 255)`, so gets the `TY_(HTMLVersion)`, at this time is `0x00020000`, which is `HT50`, 131072u, and does a `TY_(EntityName)(code,vers)` lookup...

**BUT** it encounters the code 8364, and this does **not** have `HT50`, so does **not** return the name `euro`, and breaks from the search loop, with the comment `break; /* Found code. Stop search. */`...

This shows the first problem. The expanded entity list can have many codes **more than once**, so should this `break;` be there?

You can `see` the output [tempallents1.html](http://geoffair.org/tmp/tempallents1.html), which is listed in **numeric** order, shows 2030 unique entity names, but only 1444 values, and that table is **missing** 2 entities. So quite a number of entities share the **same** value...

The second is that in the table the `euro` has a version `VERS_FROM40`. One would think that idea **from-40** would include `HT50`, so this brings into question the define `#define VERS_FROM40 (VERS_HTML40|VERS_XHTML11|VERS_BASIC)`! 

Should not `VERS_FROM40` include `VERS_HTML5`? And maybe also `VERS_FROM32`, but it seems it is **not** used anywhere, so not a problem...

And although not yet fully traced case 437468 can see these 2 apparent **problems** might effect it also...

Logically, the two entities output in **new** tidy, `&#8220;` and `&#8221;` are values that are in the table more than once, so the `lookup` would find `&OpenCurlyDoubleQuote;` and `&CloseCurlyDoubleQuote;` first, and these are marked `VERS_CHECK` == `VERS_HTML5`, so would **not** find `VERS_HTML40`, the doctype, and exit the find loop, with no `name`, thus outputs the numeric entity...

**Proposed:**

 1. Version `VERS_FROM40` have `VERS_HTML5` added, which includes `HT50` and `XH50`.
 2. That `break;` be removed, to continue to search for other matching entities.
 3. Version `HT50` be added to `VERS_FROM32`, but not important as not yet used.

The 2. could still break if a `version` **and** `name` found. Or at the time cost of always searching the **full** 2032 list, not break, to always return the lowercase form, which will always occur after the uppercase item in the current alphabetic list. 

Think I would opt for the former, but can see slight reasons for the latter...
 
So it is sort of **lucky** that test case 412 included some utf-8, as this helped isolate this problem, but of course it should be removed from a `latin1` document...

Running out of time today to effect these changes, and do a new round of tests, but will get to it soonest... others should feel free to jump in and test also... thanks...

Who would have thought that a simple table extension could lead to so much, but that's the way it goes, especially with tidy, where one action can have all sorts of unenvisaged consequences... but tidy will be better with this extended table I think...

PS: After effecting changes, found time to run a quick **regression** test, with diff below, but...

```diff
diff --git a/src/entities.c b/src/entities.c
index abba708..c1b4453 100644
--- a/src/entities.c
+++ b/src/entities.c
@@ -2175,9 +2175,11 @@ ctmbstr TY_(EntityName)( uint ch, uint versions )
     {
         if ( ep->code == ch )
         {
-            if ( ep->versions & versions )
+            if (ep->versions & versions)
+            {
                 entnam = ep->name;
-            break; /* Found code. Stop search. */
+                break; /* Issue #643 - Found code and version. Stop search? */
+            }
         }
     }
     return entnam;
diff --git a/src/lexer.h b/src/lexer.h
index 1049c48..50fa2cd 100644
--- a/src/lexer.h
+++ b/src/lexer.h
@@ -207,8 +207,8 @@ typedef enum
 #define VERS_IFRAME        (VERS_HTML40_LOOSE|VERS_FRAMESET)
 #define VERS_LOOSE         (VERS_HTML20|VERS_HTML32|VERS_IFRAME)
 #define VERS_EVENTS        (VERS_HTML40|VERS_XHTML11)
-#define VERS_FROM32        (VERS_HTML32|VERS_HTML40)
-#define VERS_FROM40        (VERS_HTML40|VERS_XHTML11|VERS_BASIC)
+#define VERS_FROM32        (VERS_HTML32|VERS_HTML40|HT50)
+#define VERS_FROM40        (VERS_HTML40|VERS_XHTML11|VERS_BASIC|VERS_HTML5)
 #define VERS_XHTML         (X10S|X10T|X10F|XH11|XB10|XH50)
 
 /* strict */
```

Now have two more to check - 

 1. 378b, a `--fix-uri no` case, now adds warning, but looks correct?
 2. 445557, comment says original `-`, to `&ndash;`, to `&#150;`?, but to me output looks better!
 
But now 437468 has dropped off - phew! Of course 412, with `utf-8`, remains... be back soonest...

**Comment by geoffmcl** [comment]
Have now pushed the above changes to the `issue-643`, including a small debug only trap change... and tested the best I can...

**Regression test differences**

**378b**

Really puzzled about the `warning` change in the **378b** output... this needs to be investigated fully...

In general however, the text output does point the user to a maybe **outdated** [International](https://www.w3.org/International/O-URL-and-ident.html) link... given that I can read considerable effort to support `international` URIs elsewhere, but this could/should be a separate issue...

After further investigation, if nothing is found **wrong** maybe the **tests** should be updated...

**445557**

Concerning test 445557, the `&#150` is now output as `&ndash;`, which seems **correct** given the test is run with the `default` config...

But am a little confused by the **comment** that it seems the original did use a dash, `-`. Why was this changed to `&#150`? But that seems to have happened **long** ago, in that my copy of `in_445557.html` also has this comment, and uses `&#150`...

Although there does appear a beginning blank line difference in a browser view test, the remainder is the **same** view. That is the **purpose** of the test seems to be met, although I seem unable to find the original 445557 in SF, to check exactly what was the problem addressed...

Certainly think the tests should be updated to match...

**412**

As already stated, think the tests should be updated removing this **utf-8** from a `latin1` test file...

**Conclusion**

In general, think PR #644 can now be merged to `next`, including an update of `tests`... baring any other comments to the contrary, will do this shortly... thanks...

**Comment by balthisar** [comment]
@geoffmcl, go for it!

**Comment by geoffmcl** [comment]
@balthisar as usual about this time was preparing the following ;=)) So not quite ready to **go for it** yet, but thanks for the encouragement...

**Question:** Why has test **378b** changed? It does **not** involve any entities...

This is a simple test, to test that `--fix-uri no` squelches the warning for HTML5 documents, so the test now **FAILS**! Why?

Well, it stems from **adding** `VERS_HTML5` into `VERS_FROM40`, and that the `TY_(HTMLVersion)(doc)` function, uses this MACRO...

Many times in the past I have tried very hard to fully **understand** this `HTMLVersion` function, and have never completely succeeded... but maybe I am getting there...

Now, in simple terms, in this case, on this document, this function **now** returns 0x100, which is `X10S`, when called from `TY_(CheckUrl)`! Since this does **not** include `VERS_HTML5`, the `ILLEGAL_URI_REFERENCE` warning will be issued.

That function will again be called in `SetXHTMLDocType`, and `ReportMarkupVersion`, which will give us `Info: Document content looks like XHTML 1.0 Strict`... corresponding to the `X10S`... but is it `XHTML 1.0 Strict`?

Previously, that function would return 0x40000, which is `XH50`, and since that is included in `VERS_HTML5`, the `ILLEGAL_URI_REFERENCE` will **not** be issued, as part of issue #378, and the reported markup will be `Info: Document content looks like XHTML5`...

So, in essence, the `HTMLVersion` function has to be modified back to what it was!

It uses `VERS_FROM40` to set a `Bool html4`, when what it really means, needs, is something like a `VERS_HTML40+XH11+XB10`, which is more or less what `VERS_FROM40` **was** before this latest change...

**How to fix this?**

`VERS_FROM40` is used some 155 times in the `entities.c` table, where I think its **new** value is very correct, and only twice in `lexer.c`.

How about the `lexer.c` uses say its own new `VERS_HTML40PX`, meaning HTML4 in all its forms, plus X, like -

```
#define VERS_HMTL40PX        (VERS_HTML40|VERS_XHTML11|VERS_BASIC)
or perhaps
#define VERS_HMTL40PX        (VERS_FROM40&~VERS_HTML5)
```

Applied a patch -

```diff
diff --git a/src/lexer.c b/src/lexer.c
index 56795cf..857017b 100644
--- a/src/lexer.c
+++ b/src/lexer.c
@@ -236,6 +236,14 @@ static struct _doctypes
   {  0,    0, 0,  no,  NULL,                     NULL,                                     NULL                                                        }
 };
 
+/* 
+ * Issue #643 - Since VERS_FROM40 was extended to include VERS_HTML5
+ * to be used in the expanded entity table some 155 times,
+ * need a special macro here to denote just HTML 4 plus XHTML,
+ * which is actually the former define of VERS_FROM40
+ */
+#define VERS_HMTL40PX        (VERS_HTML40|VERS_XHTML11|VERS_BASIC)
+
 int TY_(HTMLVersion)(TidyDocImpl* doc)
 {
     uint i;
@@ -247,7 +255,7 @@ int TY_(HTMLVersion)(TidyDocImpl* doc)
     Bool xhtml = (cfgBool(doc, TidyXmlOut) || doc->lexer->isvoyager) &&
                  !cfgBool(doc, TidyHtmlOut);
     Bool html4 = ((dtmode == TidyDoctypeStrict) || (dtmode == TidyDoctypeLoose) ||
-                  (VERS_FROM40 & dtver) ? yes : no);
+                  (VERS_HMTL40PX & dtver) ? yes : no);
     Bool html5 = (!html4 && ((dtmode == TidyDoctypeAuto) ||
                   (dtmode == TidyDoctypeHtml5)) ? yes : no);
 
@@ -262,7 +270,7 @@ int TY_(HTMLVersion)(TidyDocImpl* doc)
     for (i = 0; W3C_Doctypes[i].name; ++i)
     {
         if ((xhtml && !(VERS_XHTML & W3C_Doctypes[i].vers)) ||
-            (html4 && !(VERS_FROM40 & W3C_Doctypes[i].vers)))
+            (html4 && !(VERS_HMTL40PX & W3C_Doctypes[i].vers)))
             continue;
 
         if (vers & W3C_Doctypes[i].vers &&
```

Now this test **378b** PASSES, with flying colors... will push this change...

Moving onto test **445557**, which still intrigues me... something is **not** right here... be back soonest...

**Comment by geoffmcl** [comment]
**Question:** What exactly is the problem with test **445557**?

First looking at the `case-445557.html` input. There is a confusing comment at the top of this file -

```
<!-- the original document had – - I changed that to &ndash; - then &#150; -->
```

That first `dash` is actually `0x96`, decimal 150, which is an `EN DASH` in the 1252 Windows Latin 1 (ANSI) code page. 

Now that might be ok, in that in the tests, that input uses the default config, which contains `char-encoding: latin1`...

When tidy encounters that `0x96`, it uses `DecodeWin1252` to get the unicode 0x2013, 8211, to store in the lexer, and output a warning, `INVALID_SGML_CHARS`, `line 2 column 32 - Warning: replacing invalid character code 150`... But as you can read later still concerned how this is **output** to the html file...

Now later in the document it will encounter `&#150;`, in `ParseEntity`, and pass to `EntityInfo` to get back `0x96`, but since this is in the range of `(ch >= 128 && ch <= 159)`, ie `/* invalid numeric character reference */`, again use `DecodeWin1252`, get the unicode 0x2013, 8211, and will output an `INVALID_NCR` warning `line 8 column 12 - Warning: replacing invalid numeric character reference 150`... Note, this is a **different** warning to the first, in the comment...

So is that what this test is about? Replacing invalid numeric character references? Me thinks **not**! And the `<title>` suggests otherwise - `Convert Symbol font chars to Unicode` - and then contains `&#34;`, `&#36;`, `&#174;`, `&#217;` and `&#218;`... 

Sooooo, why all this kerfuffle about 150 and a `dash`? And is it coincidental that the replacement character for 150 is 8211, or `&ndash;`... This seems to distract from the real purpose... all of which seems to work fine...

Of course, the situation is not helped by the fact that there is no `doctype` given, so current tidy assumes html 5, while back when the test was added, tidy would be defaulting to a legacy type document...

And until this issue **change**, the `&ndash;` entity did not have a version that included `VERS_HTML5`, so tidy could only output the numeric entity `&#8211;`. Now that bit has been added to the table tidy 5.5.77.I643 does find and outputs the `&ndash;`. 

This new action could be overridden by adding `numeric-entities: yes` to the config...

Thus the **445557** test can be fixed by -

 1. Add a config file which includes `numeric-entities: yes`
 2. Accepting the new `&ndash;`, and update the expects accordingly

Decided on 1., and adding a config of `case-445557.conf`, containing -

```
// HTML Tidy configuration - 20171117
indent: auto
char-encoding: latin1
tidy-mark: no
clean: yes
logical-emphasis: yes
indent-attributes: yes
numeric-entities: yes
```

and this test **PASSES**!

But I am still concerned about 1 small thing in the html output... That is in the html output, the first substitution character in the **comment** is output as `0x13`... that does **not** feel right? So want to explore why?

But this is not really related to this issue, since that is the value also added by the current tidy... maybe should raise it as a separate issue... left as is for the moment...

Anyway, have pushed an `issue-643` branch to the test repo, with changes -

 1. testbase/case-412.html - removed utf-8, and put a `-`
 2. testbase-expects/case-412.html - new html output for the above
 3. testbase-expects/case-412.txt - new text output for the above
 4. testbase/case-445557.conf - added the above new config for this test
 5. cases/_version.txt - bump version to 5.5.78
 
Thus checking out `issue-643` here and build, and `issue-643` in tests should give a **clean** regression test run... hope other get a chance to confirm... thanks...

**Comment by balthisar** [comment]
@geoffmcl, which testbase are you using? When I run the tests on macOS using branch `issue-643` for the executable, and using the `next` branch for testing, I only see two regressions:

~~~
diff -ua ../cases/testbase-expects/case-412.html ../cases/testbase-results/case-412.html
--- ../cases/testbase-expects/case-412.html	2017-10-20 08:51:51.000000000 -0400
+++ ../cases/testbase-results/case-412.html	2017-11-17 14:11:05.000000000 -0500
@@ -1,7 +1,8 @@
 <!DOCTYPE html>
 <html>
 <head>
-  <title>Issue 412 ‚&#8364;&#8220; Warnings for Quotes</title>
+  <title>Issue 412 ‚&euro;&OpenCurlyDoubleQuote; Warnings for
+  Quotes</title>
   <style type="text/css">
   a.c1 {generic_style}
   </style>
diff -ua ../cases/testbase-expects/case-445557.html ../cases/testbase-results/case-445557.html
--- ../cases/testbase-expects/case-445557.html	2017-10-20 08:51:46.000000000 -0400
+++ ../cases/testbase-results/case-445557.html	2017-11-17 14:11:01.000000000 -0500
@@ -11,18 +11,18 @@
   <p>The predicate calculus has a number of theorems and axioms for
   proving logical statements. Here are the main symbols used in
   predicate calculus:</p>
-  <p>P(x)	&#8211; proposition &#8211; a logical statement in the
+  <p>P(x)	&ndash; proposition &ndash; a logical statement in the
   condition x.</p>
-  <p>x	&#8211; any condition <strong>in the set of possible
+  <p>x	&ndash; any condition <strong>in the set of possible
   conditions</strong>.</p>
-  <p>c	&#8211; a particular condition <strong>in the set of
+  <p>c	&ndash; a particular condition <strong>in the set of
   possible conditions</strong>.</p>
-  <p><span class="c1">"</span> 	&#8211; "For every"</p>
-  <p><span class="c1">$</span> 	&#8211; "Exists"</p>
-  <p><span class="c1">Æ</span> 	&#8211; Implication</p>
-  <p><span class="c1">Ÿ</span> 	&#8211; Conjunction (logical
+  <p><span class="c1">"</span> 	&ndash; "For every"</p>
+  <p><span class="c1">$</span> 	&ndash; "Exists"</p>
+  <p><span class="c1">Æ</span> 	&ndash; Implication</p>
+  <p><span class="c1">Ÿ</span> 	&ndash; Conjunction (logical
   and)</p>
-  <p><span class="c1">⁄</span> 	&#8211; Disjunction (logical
+  <p><span class="c1">⁄</span> 	&ndash; Disjunction (logical
   or)</p>
 </body>
 </html>
~~~ 

And in every case, the only change is the use of the entity references instead of the numeric references.

**Comment by geoffmcl** [comment]
@balthisar my turn to ask **which testbase are you using?** ;=))

I did say "have pushed an `issue-643` branch to the **test** repo, with changes -", and "checking out `issue-643` **here** and build, and `issue-643` in **tests** ...", so **both** must be on the **same** branch... not **next**...

And also see [PR 25](https://github.com/htacg/tidy-html5-tests/pull/25) in the **test** repo...

Happy re-testing, hopefully with success... thanks...

**Comment by balthisar** [comment]
I saw that, but my meaning is, with all of the changes, in _existing_ testbase, the only changes are as expected. The additional test cases in a PR, of course, should all test fine, too.

**Comment by balthisar** [comment]
@geoffmcl, okay, tested, and the issues are resolved, and I completely agree with the changes! This is a well-like contribution.

**Comment by geoffmcl** [comment]
Together with  [PR 25](https://github.com/htacg/tidy-html5-tests/pull/25), merged, in the **test** repo, seems version 5.5.85, which includes this issue, passes **all** tests, so think **this** can be closed...

Any future problems with entities should be a **new** issue... thanks...
